### ISSUE TITLE
Distributed collector configuration

### ISSUE THREAD

**Issue by swiatekm-sumo** [open-telemetry/opentelemetry-operator#1906]
> **Note**: This issue is intended to state the problem and collect use cases to anchor the design. It's neither a proposal nor even a high-level design doc.

Currently, configuration for a single Collector CR is monolithic. I'd like to explore the idea of allowing it to be defined in a distributed way, possibly by different users. It would be the operator's job to collect and assemble the disparate configuration CRs and create an equivalent collector configuration - much like how prometheus-operator creates a Prometheus configuration based on ServiceMonitors.

Prior art for similar solutions are [prometheus operator](https://github.com/prometheus-operator/prometheus-operator) with its Monitor CRs, or [logging-operator](https://github.com/kube-logging/logging-operator).

Broadly speaking, the benefits of doing this could be:
- Decoupling operational aspects of running the collector from functional aspects of configuring it. 
   Application developers could only write some piece of the configuration for their application, whereas a platform team would be responsible for running the collector.
- Allowing users to share pieces of the configuration, for example exporters which might depend on some global set of Secrets.
- Through the above two points, allowing the collector to be configured in a decentralized way, making it much easier to scale to a large amount of cluster users with different telemetry needs.

Potential problems doing this that are unique to the otel operator:
- The Otel collector does way more than either Prometheus or Fluent-Bit. Our solution should at minimum support all three signal types.
- Depending on signal type, the collector mode of operation is different. We want a DaemonSet for logs, but a StatefulSet for Prometheus metrics.
- It may be difficult to create configuration CRs which guarantee validity of the generated collector configuration, given the number of possible components.

I'd like to request that anyone who would be interested in this kind of feature, post a comment in this issue describing their use case.

**Comment by rupeshnemade** [comment]
Based on our products, I feel this would be a much needed feature. 

I have a multiple use case which are difficult to implement as of now but will be easier if OTEL has ability to have distributed configuration -
1. We need dynamic Kafka exporter configuration but as OTEL is purely static config it is very difficult to update the OTEL config dynamically based on different value of Kafka brokers.
2. Right now OTEL static config makes it tightly coupled to single set of config rules. If other team needs to add their own OTEL rule in different namespace then its not possible as there is no option of distributed config option in OTEL like Prometheus has ServiceMonitor feature with service discovery.

The proposal of distributed configuration will be very useful feature.

**Comment by wreed4** [comment]
This feature would be very advantageous to us.  As we grow as a company, it is our desire to move away from a central team needing to know about the many hundreds of other services running on our clusters.  Each team that writes a service is responsible for deploying their service and exposing any custom metrics or logs they want to pull off-cluster.  We want a central team to manage the pipeline of how those metrics and logs get pushed to our central observability platform, but we do not want the owner of that pipeline to have to know about which endpoint or which logs or which metrics should be forwarded off cluster and which should not.. or what services exist in the first place.  As stated in the initial problem statement of this issue, this is very similar to how the prometheus operator works today, and in fact that is what we use today.  In order to move to an OTEL based solution and replace prometheus as a forwarding agent, we really require this decentralization ability.

**Comment by jaronoff97** [comment]
Thanks everyone for your feedback here. I've come around to this idea and think it would be beneficial to the community @swiatekm-sumo i'm going to self assign and work on this after #1876 is complete.

**Comment by lsolovey** [comment]
I totally support this initiative and agree with use-cases already mentioned above.

Another use-case that I'd like to add is ability for developers to manage Tail Sampling configuration. We run hundreds of applications in cluster, with all observability data collected into the centralized platform. We want appliation developers to be able to configure Tail Sampling policies for their applications without touching `OpenTelemetryCollector` CRD (which contains a lot of infrastructure-related settings and is managed by the platform team).

**Comment by frzifus** [comment]
@lsolovey Could you give an example what way of configuration you would expect? Since I am working on a proposal.

**Comment by frzifus** [comment]
In summary, a good first step would be to separate the configuration of exporters from the collector configuration.
I had a conversation about this with @jaronoff97 yesterday. One possibility would be to start with a gateway and exporter CR. Here is an example of how these CRDs relate to each other.

```mermaid
graph TD;
    OpenTelemetryKafkaExporter-->OpenTelemetryExporter;
    OpenTelemetryOtlpExporter-->OpenTelemetryExporter;
    OpenTelemetryExporter-->OpenTelemetryGateway;
    OpenTelemetryGateway-->OpenTelemetryCollector;
```

Since all these CRDs are based on the OpenTelemetryCollector definition, it seems to me a requirement to support a native yaml configuration. 

- [ ] https://github.com/open-telemetry/opentelemetry-operator/issues/1707

Once this is done, we can start prototyping the gateway and exporter CRD.

- [ ] TBD

**Comment by luolong** [comment]
My attempts so far at setting up and configuring OTel Collector Operator have lead me to somewhat similar thoughts mentioned here and #1477.

The Prometheus Operator has the correct idea here, I believe.

There are basically two or three concerns here that would be useful to separate:

* Running `Operating OpenTelemetry` collector instances with all the best practice boilerplate of monitoring a Kubernetes cluster baked in.
  * Running _agent_ and _gateway_ instances
  * Provisioning _"preset"_ K8s telemetry collection (_nodestat receiver_, _k8s_cluster receiver_, _k8s_attributes processr_, etc)
  * Perhaps a UI component to visualize and edit OTel configuration (low priority).
* Collector configuration:
  * `OpentelemetryReceiver` resources for declaratively configuring telemetry sources
  * `OpenTelemetryExporter` resources for declaratively configuring telemetry destinations
  * `OpenTelemetryPipeline` resources for binding it all together
* _AutoInstrumentation_ resources

**Comment by frzifus** [comment]
I like the idea, but Ive a few open questions / thoughts:
- What would happen if `export-to-vendor` is based on a different image/version then `platform-collector`? Maybe it uses a component that does not exist in the image used by the platform collector.
- How would we handle env variable, volume, ... conflicts? We could limit this to the `platform-collector`. But this could become weird when only `export-to-vendor` requires for example specific TLS certs.
- If the `CollectorGroup`s are limited by a namespace, what is the benefit compared to a single collector configuration?

**Comment by asjgi** [comment]
Hello team,
I hope this message finds you well. I was wondering if there are any updates or progress on this feature request for distributed collector configuration? This functionality would be incredibly helpful for our use case, and I’m very interested in knowing if it’s on the roadmap or if there are any plans to address it in the near future.
Thank you for your hard work and efforts in improving this project!
Looking forward to your response.
Best regards 🙏
